Refactor worker to allow run submission to AWS Batch Service

clausman opened the pull request:
## Update March 22

After pulling in upstream master to our fork, this branch needs some work, particularly with the docker run code. This will hopefully be done soon, but for now docker runs will not work from this branch.

## Rationale
Maintaining a set of machines to act as Codalab workers under varying loads is expensive. We want an easy way to scale compute resources up and down based on load. 

AWS Batch is a service provided by AWS which allows you to submit jobs to work queues. It uses docker for encapsulating jobs much like Codalab does. By utilizing the Batch service, we can easily scale our compute resources based on load.

## Approach

In order to support Batch, all changes were made to the Codalab worker. From the Codalab master's perspective, this creates a single worker which has a large number of resources for running jobs. This Batch worker serves as a mediator between the Batch jobs and the Codalab bundles. A Batch worker maps to exactly one Batch job queue.

## Implementation

To facilitate a new type of worker, an abstraction was introduced around Run called the RunManager. A RunManager is in charge of various things involving Runs, but most importantly it is the point of truth for how to create a Run. The Worker class does not contain any information about where the Runs are being executed, it only knows how to interact with them via the RunManager and Run interfaces.

Two RunManagers exist. First the DockerRunManager which wraps the existing functionality in a Run, refactoring it as little as possible to make the changes easier to review. Next the AwsBatchRunManager which creates AwsBatchRuns which are wrappers for the Batch jobs.

## Using a AWS Batch worker
There are some pretty strong requirements needed to use an AWS Batch worker. This is because we must have a way to transfer dependency data to the compute machines. We do this using EFS, and the AwsBatchRunManager docstring outlines the requirements for this to work. 

For our internal Codalab environment, we already are using a shared file system which is mounted on the AWS Batch Compute Environment nodes. This means it can behave exactly like the existing code does and it does not need to download/upload bundle data.

## Known bugs

1) ~Folders aren't cleaned up after run finishes.~
2) ~With shared-file-system, a killed worker won't always pickup where it left off.~
3) ~`cl kill ...` doesn't seem to work as expected~
4) ~after most recent refactors, there seems to be a bug where it isn't finalizing jobs correctly~
5) if we lose a spot instance mid-run, we really should move the run bundle back to a state where it is scheduled again
6) ~Probably due to things running as root on Batch, the following error sometimes happens at when the worker attempts to finalize the run bundle `IOError: [Errno 13] Permission denied: '.../codalab-home/partitions/default/bundles/0x84622bc664c14a79b138f6b6e3dca948/.bash_history'`. I haven't looked into the cause yet though.~

jder committed:
Use docker library to allow standard authentication mechanisms for pulls (#1)

fabeschan committed:
Increment version

jder committed:
Merge remote-tracking branch 'upstream/master'

maxwang7 committed:
Merge branch 'master' into staging

maxwang7 committed:
increments version to 0.2.17

maxwang7 committed:
Merge branch 'master' into staging

jder committed:
Use docker library to allow standard authentication mechanisms for pulls (#1) (#810)

fabeschan committed:
Merge commit 'c6a0c20c2017f41072049d241ef595027719f5e3' into release

percyliang committed:
Merge branch 'master' of github.com:codalab/codalab-cli

percyliang committed:
Merge branch 'master' of github.com:codalab/codalab-cli

percyliang committed:
increment version

percyliang committed:
merge

clausman committed:
wip

clausman committed:
wip mostly working

clausman committed:
job submission working. need to support dependencies.

clausman committed:
fix log statement

clausman committed:
fix dependency

clausman committed:
fixed various things. got training job running

clausman committed:
set bundle as running

clausman committed:
refactor run and implement for aws batch run

clausman committed:
fix but

clausman committed:
fix deps showing in codalab ui

clausman committed:
mild cleanup

clausman committed:
make batch pretend to have more ram

clausman committed:
increase batch default memory. fix bug with error messages.

clausman committed:
added queue_name parameter and implemented kill

clausman committed:
refactor docker run to user run manager paradigm

clausman committed:
fix some bugs. local docker runs seems to be working

clausman committed:
switch docker run to use filesystem helper

clausman committed:
fix bug

clausman committed:
cleanup

clausman committed:
got not shared_file_system working with batch assuming shared efs

clausman committed:
added docstring to aws batch run manager

clausman committed:
remove unused change

clausman committed:
remove unused function

clausman committed:
remove unused input and output from fsm

clausman committed:
remove out of date comment

clausman committed:
Merge remote-tracking branch 'upstream/master'

clausman committed:
clean boto3 dependency so it is only done lazily in main.py

jder commented on worker/codalabworker/run_manager.py:
```diff
@@ -0,0 +1,302 @@
+        # TODO Both implement this and add setup/cleanup hooks to RunBase and use them to call these
+        pass
+
+    def path_for_dependency(self, dep):
```
Maybe a name which makes the side-effectful nature clearer? Also maybe leading underscore?

clausman committed:
remove version hack now that our internal codalab is upgraded

jder commented on worker/codalabworker/main.py:
```diff
@@ -93,10 +97,39 @@ def main():
+            try:
+                import boto3
+            except ImportError:
+                print("Missing dependencies, please install boto3 to enable AWS support.")
```
Use logging.exception?

jder commented on worker/codalabworker/worker_state_manager.py:
```diff
@@ -17,7 +17,8 @@ class WorkerStateManager(object):
 
-    def __init__(self, work_dir, shared_file_system=False):
+    def __init__(self, work_dir, run_serialize_func, shared_file_system=False):
+        self._run_serialize_func = run_serialize_func
```
Maybe cleaner to just pass in and store the run manager here and have this call `serialize` and `deserialize` as needed below rather than passing in the `serialize` function here and the `deserialize` function in `resume_previous_runs` below? (Or at least pass them both into `__init__` or both into the functions that need them; having it asymmetric suggests there's some difference but I don't see one.)

jder commented on worker/codalabworker/worker.py:
```diff
@@ -166,6 +149,7 @@ def _checkin(self):
             elif type == 'kill':
                 self._kill(response['uuid'])
             elif type == 'upgrade':
+                logging.error('ignoring upgrade message')
```
rm?

jder commented on worker/codalabworker/run.py:
```diff
@@ -336,81 +381,10 @@ def read_run_missing(bundle_service, worker, socket_id):
         }
         bundle_service.reply(worker.id, socket_id, message)
```
read_run_missing is now unused, I think?

jder commented on worker/codalabworker/run_manager.py:
```diff
@@ -0,0 +1,302 @@
+        pass
+
+
+class RunBase(object):
```
Seems a little strange this is in run_manager.py instead of run.py?

clausman replied:
Yeah, it was only because I didn't want to rename the classes too much initially. 
Really it should be something like `run.py` => `docker_run.py` and `run_manager.py` => `run.py`.

I will do that rename now that it is unlikely I will get any more conflicts.

jder commented on worker/codalabworker/worker.py:
```diff
@@ -242,13 +230,22 @@ def remove_dependency(self, parent_uuid, parent_path, uuid):
+                'error_code': httplib.INTERNAL_SERVER_ERROR,
+                'error_message': 'Read failed for unknown reason.',
+            }
+            socket.reply(message)
```
It seems a little weird to send something to the socket after you've already passed it to someone else to send data to, since they could presumably fail after sending some data (e.g. possibly an error message). Can we just remove the return value from run.read (since it's always True right now anyway) and say it's up to the implementer of run.read to send either some data or an error message?

clausman replied:
I agree. I'm just going to remove this and document this in `read`

jder commented on worker/codalabworker/run_manager.py:
```diff
@@ -0,0 +1,302 @@
+        """
+        raise NotImplementedError
+
+    def kill(self):
```
Looks like the `Run` class is still calling this internally and passing a message to it in `_check_and_report_resource_utilization`, so seems like we should add this arg back in. (Unless there's a reason I'm missing that it was removed.)

jder commented on worker/codalabworker/docker_image_manager.py:
```diff
@@ -70,6 +70,8 @@ def touch_image(self, digest):
 
     def start_cleanup_thread(self):
+        if self._max_images_bytes is None:
+            return
```
How about a quick comment saying that None means infinite room, so no need to cleanup?

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+
+    It also has the added constraint for the shared file system option, that the file system must be shared properly
+    on the compute cluster which is serving the Batch queue.
+    For an article on how to achieve this, see:
```
Missing link

jder commented on worker/codalabworker/fsm.py:
```diff
@@ -0,0 +1,64 @@
+    def update(self):
+        """
+        The update method which is called as part of the primary run loop of a FSM.
+        :return: a tuple of a new state for the FSM and a list of output messages
```
Comment is out of date; you just return the new state. 

I'm not sure about transition_noop and transition_done; it seems like given that you're just returning a state from update, `return self` is reasonably clear. I agree that `return None` isn't great, but maybe `return State.DONE` (which is None) is better? I'm just suggesting this because on first read I didn't understand whether transition_noop and transition_done were intended to be overridden by subclasses or called to cause a transition or what. (Their names made it sound to me like they could be callbacks to the state subclass for when certain transitions happened.) 

Generally might want to just add some high-level context about how to use this to the class docstring, or maybe some tests would help to demo it.

jder replied:
(Comment still out of date)

jder commented on worker/codalabworker/fsm.py:
```diff
@@ -0,0 +1,64 @@
+        self._sleep_time = sleep_time
+        # The thread for running this FSM
+        self._thread = threading.Thread(target=ThreadedFiniteStateMachine._run, args=[self])
+        self._should_run = True
```
I notice elsewhere in the code base we're using locks around shared booleans like this. I don't think we need to with CPython; does codalab target other implementations?

clausman replied:
@percyliang do you know the answer to this?

jder reviewed (changes requested):
Phew! Ok, I have read all of it except for the meat in aws_batch.py. I have to run but here's some initial feedback. I'll try to finish up the rest tomorrow.

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+    4) The Job Queue must use a Compute Environment whose AMI has a shared filesystem mounted (e.g. AWS EFS)
+    5) The shared filesystem must also be mounted on the worker machine
+    6) The absolute path to the shared filesystem must be the same on the worker and Compute Environment machines
+    7) The workers work-dir, and hence bundle_paths, must be on this shared filesystem
```
A few comments:

Maybe give the batch machines a name you can use throughout this comment? "Compute machines" or something? Then maybe just a comment along the lines of your PR description about how there's a worker machine which runs this code and mediates between codalab and the compute machines via Batch?

For (2), maybe a link to docs about setting up credentials?

For (3), maybe "must exist with the name _passed in as_ `queue_name`"?

I don't think I understand the comment about the shared filesystem. So we always need a shared FS between the worker and compute machines, but if you are using --shared-file-system then additionally that filesystem should be the same one mounted by the bundle service? (So 5-7 apply in both cases?)

clausman replied:
The `--shared-file-system` wording I had was confusing. I've updated it to hopefully be clearer. What you described about needing the same shared filesystem mounted by all things is correct.

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+        else:
+            # Set up a directory to store the bundle.
+            remove_path(self._bundle_path)
+            os.mkdir(self._bundle_path)
```
Maybe a TODO about deduplicating this with Run?

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+        if not self._bundle_service.resume_bundle(self._worker.id, self._uuid, start_message):
+            return False
+        # TODO Do we need to do anything special here? We already deserialized to the correct state presumably
+        self.create_fsm().start()
```
RE: the TODO. It looks OK to me; maybe worth just mentioning that the initial state knows how to resume by checking the bundle metadata? (Or include that in the docstring of create_fsm?)

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+        """
+        start_message = {
+            'hostname': socket.gethostname(),
+            'start_time': current_time(),
```
In the Run implementation, this uses the saved start time from when the run actually started (saved in `start`, then serialized & deserialized) rather than the current time. Should we do that here, too?

clausman replied:
I actually don't think it makes sense here since the `time` a bundle runs for batch is calculated via the batch job not via the time the worker 'started'. I'm just going to remove this.

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+        Report that the bundle is running. We note the start time here for
+        accurate accounting of time used, since the clock on the bundle
+        service and on the worker could be different.
+        """
```
This docstring is copy-pasted from `start` and I think is not really meaningful for `resume` (or `Run.resume`, which is where this came from)

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+    return {
+        'name': name,
+        'payload': payload
+    }
```
This appears to be unused.

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+
+    @property
+    def docker_image(self):
+        return self._bundle['metadata']['docker_image']
```
This property appears to be unused. (And I think this should be pulled from `resources` anyway?)

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+            '(%s) >stdout 2>stderr' % self._bundle['command'],
+            ]
+
+        return ['bash', '-c', '; '.join(bash_commands)]
```
Sad this is all duplicated from docker_client. Maybe a TODO again as a warning for folks that might try to change this?

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+    def create_fsm(self):
+        assert self._fsm is None, "FSM was already created."
+        # TODO Can this be replaced by just mounting dependencies directly?
+        dependencies = self.setup_dependencies()
```
Seems a little weird this is in `create_fsm`. It looks like `Run` only calls this in the `start` case, not the `resume` one; should we just move this into `start`? Or if we want it in both cases, maybe rename this to `drive_batch_job` and call `self._fsm.start()` at the end or something similar?

clausman replied:
This has been moved the run's `pre_start` method which the worker calls on newly created workers but not resuming ones.

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+            ))
+
+        job_definition = {
+            'jobDefinitionName': bundle['uuid'],
```
Should this be prefixed with "codalab-worker-" or something? Also maybe made not to collide if this bundle was attempted before?

clausman replied:
Names aren't unique, so I don't think collisions are a problem. I personally find the bundle uuid easiest because then you can copy-paste into codalab interfaces w/o any changes.

jder replied:
I kinda feel like being able to tell they are from codalab would be good in case we end up making Batch jobs for other things, too. (So they at least all sort together in the UI.) Tags would be better but that seems to not be a thing.

clausman replied:
Jobs are always associated with the queue, and imo this is the job of the queue to list this association. e.g. name a queue `codalab-general` and only submit Codalab things to it. Given that queues have a many-to-one relationship with compute environments, it would be just plain odd to have a queue you submit both codalab jobs to and other jobs.

clausman replied:
Also, if we have other processes which end up submitting jobs which visually looks like Codalab uuids, I will be very very sad.

jder replied:
Jobs are associated with a queue, but this is job definitions, which are global to our AWS account, so I think namespacing of some kind is important.

jder replied:
(Also are you sure they're not unique by name?)

clausman replied:
This relates to my updated response below. The more I thought about it, the more I am convinced that the job ids should go on the bundle and should be treated essentially as having 1-per-run-bundle. It seems to me that this is how Codalab currently models runs.

clausman replied:
When you register a job definition for a name which already exists, it creates a new revision and returns your the jobDefinitionArn for that revision. This will be something like `...:job-definition/0x25b12f80c09548f1a72358d75ea89ddd:2` for Revision 2 and bundle `0x25b12f80c09548f1a72358d75ea89ddd`

clausman replied:
p.s. I was also not reading the location in the code properly. I'm okay with renaming the job definitions to have a prefix. I misinterpreted this comment to be about the job itself.

If all you care about is the job definition, and you feel strongly, then I am happy to change. (just give me a 👍 )

jder replied:
ok, yeah, that makes more sense. I think jobs can be whatever but I do think we should namespace the job definition, thanks.

clausman replied:
After more reflection, I decided my objections were silly and I should just listen to your wisdom in the first place. I've changed the naming convention for both job definitions and runs to `codalab-worker-$VERSION-$UUID`

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+                'vcpus': cpus,
+                'memory': int(memory_bytes / BYTES_PER_MEGABYTE),
+                'command': self.docker_command,
+                'volumes': map(lambda pair: pair[0], volumes_and_mounts),
```
`[vol for vol, _ in volumes_and_mounts]` might be more pythonic. (Same below.)

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+
+    """
+    Waiting for Batch to schedule the job
+    """
```
Move to start of class definition? Also it seems like this is not just waiting for it to schedule but also while the job is running?

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+        if self.should_refresh() or self._job is None:
+            response = self._batch_client.describe_jobs(jobs=[self._job_id])
+            self._job = response['jobs'][0]
+        return self._job
```
Why do we want to fetch every 30 seconds instead of every 5? (Which would be the result if we just fetched directly in Running, right?) If we do want it to be slower, maybe just expose an `update_period` member on `State`?

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+class JobFetcher(object):
+    def __init__(self, job_id, batch_client):
+        self._last_check_time = 0
+        self._check_frequency = 30  # seconds
```
How about `self._check_period`? (Could also use a timedelta here instead of int seconds and use datetime directly.)

clausman replied:
this was rmed

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+
+        now = current_time()
+        started = job.get('startedAt')
+        runtime = now - (started / 1000) if started else 0
```
Might be nice to use datetime & timedelta here, too, and just convert when passing it to update_metadata

clausman replied:
I decided not to do this, because the conversion to and from everywhere seemed like it would be even more code. Really this change should be made to allow datetime to be used rather than epoch time, and until then I think it will be more trouble than it is worth to use the proper types here.

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+    def update(self):
+        job = self._job_fetcher.fetch_job()
+        status = job['status']
+        # The contain contains information about the most recent docker container used to run the job
```
typo `contain` -> `container`

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+        if status == BatchStatus.Succeeded:
+            run_status = 'Succeeded'
+            finalize_message = {
+                'exitcode': container.get('exitCode'),
```
Should this be `.get('exitCode', 0)` to match the Failed case? Or maybe they should both default to None?

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+        # Update the bundle locally
+        self._bundle['metadata'].update(kwargs)
+        # Update the bundle on the bundle service
+        self._bundle_service.update_bundle_metadata(self._worker.id, self._bundle['uuid'], kwargs)
```
The `Run` code avoids blowing up if the bundle service throws an exception here. Should we replicate that behavior?

jder commented on worker/codalabworker/fsm.py:
```diff
@@ -0,0 +1,64 @@
+    """
+    A FSM which will run on a separate thread.
+    """
+    def __init__(self, initial_state, sleep_time=0.1):
```
Maybe document units of sleep_time?

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+
+class Submit(AwsBatchRunState):
+    def update(self):
+        job_definition = self._bundle['metadata']['batch_job_definition']
```
use `self.metadata` (and other places below)?

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+        return run
+
+    def serialize(self, run):
+        assert isinstance(run, AwsBatchRun), "Could not serialize run which was from different run manager."
```
Should we have this same assert in the other RunManager?

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -0,0 +1,539 @@
+        started = job.get('startedAt')
+        runtime = now - (started / 1000) if started else 0
+
+        run_status = 'Batch Status: %s' % status
```
Should we also include the statusReason (if there is one)?

jder reviewed (changes requested):
Ok, I think that's it from me. Awesome stuff overall. 

One high-level question: what's the logic for storing the batch IDs in the bundle metadata rather than in the Run? On the face of it, the Batch job id & definition are really a property of the run, not of the bundle (since it's only applicable to a particular run manager, for example). Just thinking that if there's a problem starting the job and codalab re-stages it, we'd probably want to, say, make a new job definition with a unique name rather than reusing the one in the bundle metadata.

clausman commented:
@jder wrt the batch job ids being put in the bundle metadata, I agree with you. Primarily, it stems back to wanting to be able to 'continue' a job where I left off when debugging. Since most development was done using a `--share-file-system` worker (which don't save state), putting it in the bundle metadata made this possible. I agree though, that it would really make more sense in the worker state and then it is up to each worker to continue from its previous state just like docker workers do.

*[Update]*

Actually, I think I am changing my mind on what I said above. Conceptually on Codalab, a run bundle is only ever run exactly once. So if a run bundle has been submitted to Batch at all, we should always respect w/e the output of that batch job id was. If there is a problem with a run, we should propagate whatever that problem was to Codalab. 

The only case I'm aware of where Codalab restages bundles is if a worker goes offline, and in this case you really do want to re-use the batch job if it is there since just because the worker went offline doesn't mean Batch isn't still running the job. 

I suppose this doesn't gracefully handle the case where the Batch worker who picks up the restaged job doesn't have access to the already running Batch job, but to me it seems like that should be reported as a failure rather than re-running (a potentially expensive) job which has already been run.

clausman committed:
first batch of jders comments

clausman committed:
jders second batch of comments

clausman committed:
fix refactor bug

clausman committed:
fix refactor bug

clausman committed:
fix refactor bug

clausman committed:
make run loop faster for batch state machine

clausman committed:
move run to docker_run and run_manager to run

jder commented:
Re: resume/worker_offline/job in metadata, I think it's fine to leave it the way it is now and maybe just get Percy's 2 cents.

jder committed:
Make Content-Disposition headers include "attachment" (#4)

The previous version was technically in violation of the RFC,
so we have an HTTP client that was complaining about it.

clausman committed:
fix bug causing vcpus to not set properly

clausman committed:
add tests for fsm and add synchronous version

clausman committed:
create some initial tests for aws run manager

clausman committed:
first pass of batch and fsm tests

clausman committed:
better mocking

clausman committed:
remove dependency folders and update tests

clausman committed:
reformat

clausman committed:
add pre_start post_stop hooks and use for dependencies

clausman committed:
add filesystem monitoring

clausman committed:
fix bug in fs monitor setup

clausman committed:
remove redundant folder cleanup

clausman committed:
save worker state on shared filesystem

clausman committed:
move upload to filesystem_run

clausman committed:
fix tests and bug in dependency cleanup

clausman committed:
Merge branch 'master' into aws_batch

clausman committed:
fix start_time bug

clausman committed:
better naming for batch jobs

clausman committed:
fix duplicate call of setup_dependencies

clausman committed:
fix bad kill call

clausman committed:
add pre_start call that somehow went missing...

clausman commented:
@jder RFAL when you are ready (no rush). All tests are passing for me locally.

Here is an overview of the changes since your last look since there were quite a few:
1) Add tests for aws batch and a few others
2) Add pre_start and post_stop hooks to Run to allow before/after setup tasks
3) Use these new hooks in FilesystemRunMixin to setup and cleanup dependencies
4) Allow shared-file-system to save state (Percy approved this) which lets us pickup batch jobs where we left off if the worker dies
5) Add filesystem monitoring to batch jobs so they submit the data_utilization while running
6) Fix all known bugs and make changes from PR comments

clausman committed:
install boto3 into codalab worker docker image

jder commented on tests/worker/aws_batch_test.py:
```diff
@@ -0,0 +1,299 @@
+        state, batch_client = self.create_state(status=BatchStatus.Failed)
+        next_state = state.update()
+
+        self.assertIsInstance(next_state, Cleanup, 'should continue running when batch it')
```
I don't understand the assert message (same one just below)

jder commented on tests/worker/aws_batch_test.py:
```diff
@@ -0,0 +1,299 @@
+            },
+        }]}
+        bundle_service = state._bundle_service
+        bundle_service.update_bundle_contents.assert_not_called()
```
Some of tests in this file look like they might be pretty brittle to me. They rely on a lot of implementation details of the underlying classes, sometimes several levels deep. (e.g. mocking out os.rmdir, verifying private state variables are set correctly, etc) Maybe we can talk about this offline but maybe this kind of integration test would be easier to maintain if it either higher-level (e.g. run a bundle and verify the results) or lower-level?

clausman replied:
I completely agree with this but I'm not going to fix before merging.

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -207,21 +214,14 @@ def create_fsm(self):
 class BatchStatus(object):
     """
-    Constants for the statuses a Batch job can be in.
+    Constants for the statuses a Batch job can be in.293
```
typo

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -309,39 +324,42 @@ def update(self):
         return ['bash', '-c', '; '.join(bash_commands)]
 
-    def create_job_definition(self, dependencies):
+    def get_job_definition(self):
```
I like it when tests encourage good refactors :)

jder commented on worker/codalabworker/aws_batch.py:
```diff
@@ -509,7 +528,7 @@ def update(self):
             run_status = 'Succeeded'
             finalize_message = {
-                'exitcode': container.get('exitCode'),
+                'exitcode': container.get('exitCode', 0),
```
Why default to 0? Seems like it could hide something going wrong?

clausman replied:
This is inside a match on the job status as reported by batching being "succeeded". I believe it is the case that there isn't always an exit code set however I'm not sure of that and it's been too long for me to be comfortable touching this so I'm going to leave it as-is.

jder reviewed (approved):
Generally looks great. I especially like the new tests. I have a couple comments but nothing that should stop merging when you're happy. Thanks!

clausman committed:
clean aws batch volume names

clausman committed:
always set request_memory for aws batch runs

clausman committed:
increase batch default memory for peace of mind

clausman committed:
handle duplicate names after cleaning

clausman committed:
Merge pull request #5 from semanticmachines/fix_some_bugs

Fix batch request_memory and dependency name bugs

clausman committed:
Merge branch 'master' into aws_batch

# Conflicts:
#	codalab/common.py
#	worker/codalabworker/main.py
#	worker/codalabworker/run.py
#	worker/codalabworker/worker.py
#	worker/setup.py

clausman committed:
resolved some merge conflicts, still more to go in docker land

clausman committed:
fix merge conflict bug

clausman committed:
assert docker image before submitting to batch

clausman committed:
fix bug with aws batch worker setting killed jobs to worker_offline

clausman committed:
Add error recovery mechanism to FSM so we don't always kill the threads.
Also increase hard-coded memory limit for Batch.

clausman committed:
pr comments including adding resume to Running state

clausman committed:
log error before cleanup; respect bundle start time on batch resume

clausman committed:
almost always recover in the running state

clausman committed:
Merge pull request #10 from semanticmachines/recover

Add error recovery mechanisms to FSM so we don't always kill threads.

clausman committed:
some typo fixes
